frontend-react/src/config/api/RSEndpoint.ts

```diff
@@ -0,0 +1,296 @@
+import { QueryFunctionContext } from "@tanstack/react-query";
+import axios, { AxiosRequestConfig, RawAxiosRequestHeaders } from "axios";
+import omit from "lodash.omit";
+
+import { getMembershipsFromToken } from "../../hooks/UseOktaMemberships";
+import { OKTA_AUTH } from "../../okta";
+import {
+    AppInsightsHeaders,
+    getAppInsightsHeaders,
+} from "../../TelemetryService";
+import config from "..";
+import { RSNetworkError } from "../../utils/RSNetworkError";
+
+export const RSAuthenticationTypes = {
+    OKTA: "okta",
+} as const;
+export type RSAuthenticationType =
+    (typeof RSAuthenticationTypes)[keyof typeof RSAuthenticationTypes];
+
+export interface RSRequestHeaders extends AppInsightsHeaders {
+    "authentication-type"?: RSAuthenticationType;
+    authorization?: string;
+    organization?: string;
+    [k: string]: undefined | string | number | boolean;
+}
+
+export type RSRequestConfig = Omit<
+    RequiredProps<AxiosRequestConfig, "url">,
+    "headers"
+> & { headers: RSRequestHeaders };
+
+export async function getRSRequestHeaders(
+    headers?: Partial<RSRequestHeaders>
+): Promise<RSRequestHeaders> {
+    const accessToken = (await OKTA_AUTH.tokenManager.getTokens()).accessToken;
+    // Isn't membership a mandatory thing? Should membershipsFromToken throw an error
+    // if not none found so usages of it can safely assume a non-undefined return?
+    // Assuming we'll never get an empty string for now.
+    const { activeMembership } = getMembershipsFromToken(accessToken);
+    const organization = activeMembership?.parsedName ?? "";
+
+    let apiHeaders: RSRequestHeaders = { ...getAppInsightsHeaders() };
+
+    if (accessToken) {
+        apiHeaders = {
+            ...apiHeaders,
+            "Content-Type": "application/json",
+            "authentication-type": "okta",
+            authorization: `Bearer ${accessToken.accessToken}`,
+            organization: organization,
+        };
+    }
+
+    apiHeaders = { ...apiHeaders, ...headers };
+
+    return apiHeaders;
+}
+
+const { API_ROOT } = config;
+
+// odd that there isn't already a useable implementation of this somewhere
+// or is there? where should this live?
+export const HTTPMethods = {
+    GET: "GET",
+    POST: "POST",
+    PUT: "PUT",
+    DELETE: "DELETE",
+    PATCH: "PATCH",
+} as const;
+export type HTTPMethod = (typeof HTTPMethods)[keyof typeof HTTPMethods];
+
+/**
+ * Copied from react-router for internal use with Endpoints.
+ * Please use react-router's types directly for react-router
+ * work.
+ */
+export type ParamParseSegment<Segment extends string> =
```

Comment:
I'm a bit wary of the fact that we're copying this from react-router source code.  It'd be one thing if this (and the path interpolation) were explicitly exported by react-router, but having to maintain this type separately feels like unnecessary maintenance on our end.  (It also looks like [there's been some simplification](https://github.com/remix-run/react-router/pull/9090](https://github.com/remix-run/react-router/blob/main/packages/router/utils.ts#L218)) in the latest version?)

Regarding the path interpolation, I wonder if it'd just be simpler to have all of the paths be functions that, at the time of request, could take in certain parameters:

```js
new RSEndpoint({
    path(pathArg1, pathArg2, pathArg3, params) {
       const basePath = `/api/${arg1}/${arg2}/${arg3}`;
       if (params) {
         return `${basePath}?${transformQueryParams(params)}`;
       } 

       return basePath;
    },
    ...otherStuff
});
```

`toDynamicUrl` is interesting logic, but my feeling is that it's more complicated than it needs to be because 1) we really only need an exploded string path returned and 2) missing segments can just be enforced through argument types to something like the `path` method above.  It's also specifically around dynamic path components, not dynamic query parameters -- so why not make everything dynamic?  

What are your thoughts about this?

Reply:
Agree with Stephen, can this be simplified?